Remove chown and merge chmod flags

principis opened the pull request:
#### Database Migration
NO

#### Description
Removed all `chown` operations because they never work.
I haven't touched `chmod` because I doubt the masks should be merged.

I edited both Sonarr.Api.V3 and NzbDrone.Api because I don't know if it's still used for V3, let me know if I should remove my edits.

This is my first PR here, so let me know what I need to change. :)


#### Todos
- [ ] Tests
- [ ] Documentation


#### Issues Fixed or Closed by this PR

* #3752

Taloth commented on src/NzbDrone.Mono/Disk/DiskProvider.cs:
```diff
@@ -89,11 +97,6 @@ public override void CopyPermissions(string sourcePath, string targetPath, bool
                     Syscall.chmod(targetPath, srcStat.st_mode);
                 }
-
-                if (includeOwner && (srcStat.st_uid != tgtStat.st_uid || srcStat.st_gid != tgtStat.st_gid))
```
Keep this in. I think it was added to deal with a particular edgecase when copying folders.
It'll still not be effective in much scenarios, lacking appropriate permissions, but it's not user facing.

Taloth replied:
Actually... this was the issue https://github.com/Sonarr/Sonarr/issues/3161
'Implemented' in v3 last year.

I'm torn a bit between tearing it out and leaving it in .... 😕 

... yup, tear it out.

principis commented:
I didn't see the benefit because it removes a feature and adds validation, but I understand the reasoning. I'll update the pr.

principis commented:
@Taloth I've updated the pr with the chmod changes. I've also added some tests.

I'm not sure if I should change `FileChmod` to just `Chmod` (in the code, not ui) it could be better, but I left it for now.

Taloth commented on src/NzbDrone.Common/Disk/DiskProviderBase.cs:
```diff
@@ -360,6 +361,31 @@ public void SetPermissions(string filename, WellKnownSidType accountSid, FileSys
 
+        protected static string GetFolderPermissions(string filePermissions)
+        {
+            var builder = new StringBuilder();
```
This works. But I think it's cleaner to manipulate the output of `FilePermissions fp = NativeConvert.FromOctalPermissionString(mask)` that we're already getting.
you can even just mask out the 3 'r' (see FilePermissions), shift them down 2 bits and or.

principis replied:
Sure, I'll change it.

Taloth replied:
forgot to mention that you also need to make sure the user doesn't add executable flag to filechmod. Either through validation, or masking them out later.

principis replied:
I'll improve the validation then, because currently it only checks if it's not empty.

Taloth commented on src/NzbDrone.Mono.Test/DiskProviderTests/DiskProviderFixture.cs:
```diff
@@ -151,11 +151,73 @@ public void should_copy_folder_permissions()
+            Syscall.stat(tempPath, out fileStat);
+            NativeConvert.ToOctalPermissionString(fileStat.st_mode).Should().Be("0750");
+
+            Subject.SetPermissions(tempPath, "0041");
```
Doesn't this remove owner permissions? Might need a bit of code to cleanup the test folder afterwards.

principis replied:
Yes, but it should be fine because it's a temp file and the file can be removed because the parent folder should have enough permissions.
I added that test, although it's unrealistic, to make sure the function works. I'll remove the file afterwards.

Taloth replied:
Paths retrieved via GetTempFilePath are automatically removed after the test. So if the permissions don't prevent removal then no further changes are needed.

principis commented:
Sorry for the delay, I'm quite busy with upcoming exams.

I've changed `GetFolderPermissions` to use `FilePermissions` and also added proper validation. I changed the default value of `0644` to `644` because the 0 isn't necessary and may be confusing for some people (new unix users which aren't aware of the special bits). Alongside prohibiting the executable bit, I also prohibited the Setuid, Setgid, and Sticky Bits, because I think media files should never have those set. Let me know if I'm wrong! :)

Taloth commented on src/NzbDrone.Core/Configuration/ConfigService.cs:
```diff
@@ -254,32 +254,11 @@ public bool SetPermissionsLinux
         public string FileChmod
         {
-            get { return GetValue("FileChmod", "0644"); }
+            get { return GetValue("FileChmod", "644"); }
```
Note: `0` prefix is used in some programming languages to indicate 'octal notation'. So that's likely why it was there.

principis replied:
I changed it back.

Taloth commented on src/NzbDrone.Core/Datastore/Migration/140_remove_folderchmod_config.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        protected override void MainDbUpgrade()
+        {
+            Execute.Sql("DELETE FROM Config WHERE Key = 'folderchmod'");
```
Note sure if we should merge it with filechmod. Like `filechmod = (filechmod | folderchmod) & ~(execute flags)`.
I don't expect anyone to have them configured differently, so it's probably not needed.

Taloth replied:
Also delete chown user&group

principis replied:
I think they shouldn't be merged, it may cause unexpected issues.
Sorry forgot the chown :)

Taloth commented on src/NzbDrone.Core/Sonarr.Core.csproj:
```diff
@@ -28,6 +28,9 @@
     <Reference Include="Growl.CoreLibrary">
       <HintPath>..\Libraries\Growl.CoreLibrary.dll</HintPath>
     </Reference>
+    <Reference Include="Mono.Posix">
```
No Mono.Posix in Sonarr.Core, hide that in Sonarr.Mono

Taloth commented on src/NzbDrone.Core/Validation/FileChmodValidator.cs:
```diff
@@ -0,0 +1,27 @@
+            {
+                try
+                {
+                    var permissions = NativeConvert.FromOctalPermissionString(x);
```
I understand the conundrum, the only thing I can think of is to either add a IsValidFilePermissionMask function to DiskProvider or copy the flags and do it ourselves. First option is easier I think, DiskProviderBase would simply return throw new NotSupportedException(), with the real implementation in Sonarr.Mono. 
You'd need a non-static validator, similar to pathExistsValidator, to access IDiskProvider.

principis replied:
Should be fixed now. :)

principis commented:
> A bit of a review delay, but was a bit busy at work.
> 
> Biggest issue is mono assembly leaking out of containment :)

I'm quite inexperienced with this scale of C# project, I thought it may be a problem, but didn't know how to deal with it. Thank you! :) (I actually forgot to commit the csproj in the first instance.)

I have an exam coming up, so I'll fix the requested changes on Friday.

Taloth commented:
Thank you for your effort in this PR. I realize it grew from 1 small up-for-grabs to something a bit more complicated.

Good luck with the exam.

principis commented:
No problem, I learned things and enjoyed doing it. :)

I've improved the validation by also checking if the user can read and write to the file. I think that should always be the case for Sonarr to function properly. (?)
Please let me know if I overlooked something.

Taloth commented on src/NzbDrone.Core/Datastore/Migration/140_remove_chown_and_folderchmod_config.cs:
```diff
@@ -0,0 +1,17 @@
+        {
+            Execute.Sql("DELETE FROM Config WHERE Key = 'folderchmod'");
+            Execute.Sql("DELETE FROM Config WHERE Key = 'chownuser'");
+            Execute.Sql("DELETE FROM Config WHERE KEY = 'chowngroup'");
```
Nitpick: 'KEY' -> 'Key'...
Also consider `Key IN ('folderchmod', 'chownuser', 'chowngroup')` turns it into a single query

principis replied:
Sorry typo :/

Taloth commented on src/NzbDrone.Mono/Disk/DiskProvider.cs:
```diff
@@ -72,13 +72,49 @@ public override void InheritFolderPermissions(string filename)
+            try
+            {
+                var permissions = NativeConvert.FromOctalPermissionString(mask);
+                return (permissions & (FilePermissions.S_ISGID | FilePermissions.S_ISUID | FilePermissions.S_ISVTX |
```
I had to read it twice to understand what you were really checking, so probably better to split this up.
Separately check for setuid/setuid/sticky, return false otherwise.
Separately check no execute flags, return false otherwise.
Then check for rw for user.
That should make it easier to follow.

Taloth commented on src/NzbDrone.Api/Config/MediaManagementConfigModule.cs:
```diff
@@ -1,16 +1,16 @@
         {
-            SharedValidator.RuleFor(c => c.FileChmod).NotEmpty();
-            SharedValidator.RuleFor(c => c.FolderChmod).NotEmpty();
+            SharedValidator.RuleFor(c => c.FileChmod).SetValidator(fileChmodValidator).When(c => !string.IsNullOrEmpty(c.FileChmod));
```
I was thinking about this and on windows the validator would still run and thus throw a NotSupportedException. probably better to add an OsInfo.IsMono in the When condition.

Taloth commented on src/Sonarr.Api.V3/Config/MediaManagementConfigModule.cs:
```diff
@@ -1,17 +1,17 @@
             SharedValidator.RuleFor(c => c.RecycleBinCleanupDays).GreaterThanOrEqualTo(0);
-            SharedValidator.RuleFor(c => c.FileChmod).NotEmpty();
-            SharedValidator.RuleFor(c => c.FolderChmod).NotEmpty();
+            SharedValidator.RuleFor(c => c.FileChmod).SetValidator(fileChmodValidator).When(c => !string.IsNullOrEmpty(c.FileChmod));
```
Same, OsInfo.IsMono guard.

principis commented:
Ok added the changes. I'll rebase to solve conflicts.

Taloth commented:
Thank you for your contribution.

imathew commented:
Hi all, I understand that chown wasn't working well so its removal makes sense, but it was useful to have chmod be separate for files and directories. In my case, for example, I always want directories to be 2775, and files 0664.

I would usually leave the directory perms alone (they're created with correct permissions at the OS level) and have Sonarr set the file permissions alone. I can't use this new merged setting, though, as it sets new directories to 0775 (overriding my 2775). This has flow-on effects to groups on my system, as subsequent season folders are then created with incorrect permissions, locking them out.

Thanks, appreciate your work!

Taloth commented:
> though, as it sets new directories to 0775 (overriding my 2775).

That was fixed like yesterday.

imathew commented:
> 
> 
> > though, as it sets new directories to 0775 (overriding my 2775).
> 
> That was fixed like yesterday.

Ha, sorry! I've been troubleshooting my file issues today and just figured it out, and found this. Sorry I should have searched better.

Taloth commented:
It's ok, the timing was just incredible :smile: Be sure to update Sonarr to at least 3.0.4.1016

d00nicus commented:
As someone who the chown functionality was working well for, the total removal of it is a poor outcome - I've gone from a working setup to needing to execute chown on my library on a regular basis if I want to stay on the current version.

I was under the impression the point of advanced options being hidden by default was to hide them from new/inexperienced users, a user pull request that unilaterally removes something already hidden under advanced simply because it doesn't work for *some* users (which I'm assuming is a group that @principis falls into) seems sub-optimal at best. 

For the time being I've rolled back to 3.0.3.750 (as the only Linux build I have saved from before this was merged) but I'm glad to at least have found the specific commit responsible for this so that I can look at undoing it, but it'd be nice to not have to do that every time I need/want to update Sonarr however.

markus101 commented:
Sonarr still supports changing the group, just not the user.

> because it doesn't work for some user

It only works for _some_ users, specifically users that are running Sonarr as root, which is not recommended on any system. Removing is removes the foot gun since it doesn't work in more cases than it does work. If you have a need to change the owner and you're running as root you can use a custom script to achieve the same thing.

d00nicus commented:
Scripts (or forking) it is then I guess, when I next update to a version newer than this commit anyhow - unless anything vital gets fixed that affects me, or a feature I really need is added I don't see a rush to update from 3.0.3.750 for the time being.

Thanks for responding, even if not the response I was hoping for - I appreciate it.

principis commented:
> I was under the impression the point of advanced options being hidden by default was to hide them from new/inexperienced users, a user pull request that unilaterally removes something already hidden under advanced simply because it doesn't work for _some_ users (which I'm assuming is a group that @principis falls into) seems sub-optimal at best.

I'm sorry this didn't work out as expected. I like Sonarr, wanted to contribute and saw the issue so I implemented it as requested there.
_I am confused what was wrong with chmod tho, I've implemented it such that 2775 should work fine... (I think)_

As @markus101 already said, chown only works when Sonarr is run as root, so I still stand behind removing it.

d00nicus commented:
I appreciate your position on this, although I do think at minimum any patch that alters the configured behaviour in regards to file permissions/ownership in this way without an opt-out needs to display some sort of notification in the GUI that those settings have been changed/reverted when the config is set to a non-default value rather than waiting for affected users to discover the mis-set files on their own.

I can respect *why* you did it, but I strongly believe it shouldn't have been a silent change to the running configuration.

Taloth commented:
> I like Sonarr, wanted to contribute and saw the issue so I implemented it as requested there.

@principis You did, the contribution worked exactly as it was supposed to.

We did get quite a few support questions from people not understanding what a file mask of 644 meant, and that if they wanted 755 for the folder they would need to set perms to 644 and automatically get 755 for folders. So in a later change i swapped the file-only mask to folder-only and added a lot of ui sugar to show the various masks and permissions as well as the common presets. Functionally the same, just presented differently to the user. I haven't seen support questions about it since.

> I am confused what was wrong with chmod tho

As for 2775, that was an pre-existing bug in sonarr, as far back as v2. Sonarr would remove the special bits (sticky/setgid bits) from existing folders when the mask was set to '775' preventing setgid propagation to new subfolders. One of our regular users pointed that out and I fixed it as well. In v2 some users set 644 to files and 2755 for folders to work around that, such as imathew posted above. (FYI the cmdline `chmod 0775` and `chmod 755` are functionally different since the former clears the special bits, where the later does not)
We no longer allow setting setgid bits in sonarr, but simultaneously we no longer discard it from existing folders thus properly propagating the setgid on the root folder to series->season folders. So people can now just use the cmdline to set setgid on the appropriate library folders once, and have them propagate properly as sonarr creates new folders.

I have to note that this change was made in Sonarr v3 beta, specifically because it's a potentially breaking change. And beta users are explicitly notified on the sonarr.tv website that we can and will be make breaking changes during the v3 beta phase.

@d00nicus If you're considering forking or running your own sonarr version, I strongly encourage you to first reevaluate how you're doing permissions.
Talk to us on discord if you want to clarify what your particular workflow is, we have plenty of people willing to discuss it and offer advise. It's entirely possible you have an unique workflow, but I'm sure an alternative approach is available that doesn't require you running your own sonarr version.

d00nicus commented:
> I have to note that this change was made in Sonarr v3 beta, specifically because it's a potentially breaking change. And beta users are explicitly notified on the sonarr.tv website that we can and will be make breaking changes during the v3 beta phase.

I understand (and respect the why - breaking changes happen, and they're often unavoidable) - having not had my server running for a while and recently updating it, it would just have been nice to *know* that a breaking change had altered my configuration so that I could change my workflow accordingly, the update screen in Sonarr doesn't show changelogs that far back (would definitely nice to be able to see more there), so the first I knew was when permissions issues started appearing.

I'll likely pop onto the Discord in a few days, as I'm not going to be around much over the weekend/Monday, although regardless of outcome there, as a developer forking a copy to tinker with for curiosity's sake still holds an interest

principis commented:
@Taloth Thank you for your explanation :smile:
